### ISSUE TITLE
phpcomplete - Eat your own dog food

### ISSUE THREAD

**Issue by sergeyz** [phpcomplete/phpcomplete#1]
# phpcomplete

**Goal of this tool/lib is to unify different approaches of PHP (and frameworks in advance) code completion using PHP**

Currently we have several implementations developed in parallel, we can join to create tool we like
### Known implementations

https://github.com/m2mdas/phpcomplete-extended - with Symfony/Laravel support, not maintained anymore
https://github.com/mkusher/padawan.php - Sublime/Vim support (https://github.com/nikic/PHP-Parser used) with Symfony support https://github.com/mkusher/padawan.php-symfony
https://github.com/Peekmo/atom-autocomplete-php - Atom support (Looks like own parser) with Symfony support too https://github.com/Peekmo/atom-symfony2

Subset of Vim plugins based on ctags
https://github.com/vim-php/phpctags
https://github.com/shawncplus/phpcomplete.vim

Libs implementation:
https://github.com/cweagans/theforce
https://github.com/deweller/PHPCodeIntel - looks like not maintained anymore
### Architecture

To make lib usable split existing tools in packages required
##### Packages
- root - unite packages into one tool
- engine - package with matching analysis functionality (semantic in advance)
- index - index schema and tools
- storage - store index, cache layer
- reader - read sources
- parser - parse sources to index
- cli - command line access
- server - pretty good for online IDE's
##### Providers
- vim - provider for vim
- atom - provider for atom https://github.com/atom/autocomplete-plus
- sublime, brackets
  and so on
### Challenges
- :star: fast - develop with performance in mind
- :star: extendable - events, advanced configuration
- :star: intelligent - scope analysis, suitable results
- :star: well tested/bug free
### Requirements
- caches and performance optimization
- events, advanced configuration
- implement approach similar IntelliSense
- unit/integration test cases, test coverage on high level and static code analysis in development

**Comment by sergeyz** [comment]
@m2das, @mkusher, @Peekmo, @shawncplus, @cweagans, @vim-php hey there! I would appreaciate hearing your opinion on this

**Comment by deweller** [comment]
I'd love to see this happen.  This is a need I would like filled for Sublime Text.

I'd be interested in contributing to this in my spare(?!) time.

**Comment by shawncplus** [comment]
cc'ing @complex857 on this

**Comment by cweagans** [comment]
After doing a very significant amount of research on the topic of intelligent code completion, I've concluded that Omnisharp is the model that pretty much any code intelligence engine should follow. I'd also say that one should *not* attempt to integrate with each individual editor. Instead, integrate your completion engine with `valloric/ycmd`, and then you can more or less instantly leverage the ycmd integrations that exist for many editors already.

I'd also suggest taking a look at `codeintel`, which is the code intelligence component of Komodo Edit (open source). You can do `pip install codeintel`, and then run `codeintel`, and you get the Komodo code completion engine running as a standalone daemon. Among other languages, codeintel supports PHP reasonably well. I think the sources for that are [here](https://github.com/SublimeCodeIntel/CodeIntelSources), but I'm not sure.

I likely won't have a lot of time to spend on this, but I'll follow the project and contribute where I can.

**Comment by cweagans** [comment]
Oh, one other thing: I think for something like this, a more monolithic model makes sense. That is, rather than having a plugin system that allows end-users to customize what the completer will and won't do, it should just work more like the completion engine that's built into PHPstorm. It should handle anything you throw at it without having to fiddle with a config file.

**Comment by mkusher** [comment]
@cweagans phpstorm engine has great plugin system :)

@sergeyz I'm currently working on refactoring padawan.php, because have many ideas, how I can improve type guessing, project management and index updating. Have a look at new structure I'm working on https://github.com/mkusher/padawan.php/tree/features/custom-scopes/src

What, I think, can be unified - classes and interfaces, that store information about project, classes, interfaces, methods and etc. They can be used for completion, for linter, for refactoring and so.
I event started to think that padawan should be a lib for a php-project management and other things like completer or linter should be plugins.

**Comment by sekjun9878** [comment]
I haven't tried this personally, but a "gold-standard" open source implementation of PHP completion may be available at https://github.com/Valloric/ycmd/pull/120. Of course, the aim of this project is to possibly port that functionality in to PHP code itelf, but just pointing that out that according to the author that patch is fully functional and ready to roll.

**Comment by sekjun9878** [comment]
@cweagans Doesn't PHPStorm handle Symfony2 and other CMS integrations precisely through the use of plugins though?

**Comment by cweagans** [comment]
@mkusher @sekjun9878 Yes, I suppose that's true. However, those are just enhancements. Why not build that directly into the completer and have the completer recognize when there's a symfony project or a Drupal project or something? I guess it doesn't really matter if it's through a plugin, but the thing that I'm concerned with is discoverability. It's hard enough to find a decent PHP completion engine. It's even harder to find plugins for it.

Using @mkusher's padawan as an example, the plugins listed in the wiki are very very very small chunks of code. Why not just move them into the main project? Plus, this alleviates the problem of having a million different plugins to do the same thing. Symfony support should be baked in and should "just work". If it doesn't work like it should, a developer should just submit a PR to the main project to fix it (rather than creating an entirely new project). That is, if the goal is to centralize effort on making PHP code completion work, why go through all that effort only to fragment the functionality out into a plugin ecosystem? It's just not worth it.

**Comment by sekjun9878** [comment]
@mkusher @sergeyz @cweagans A set of core included-in-default-distrubution plugins can easily solve that issue. That's more of a problem of distribution and code management rather than an architectural decision isn't it? After all, code for integrating different CMS / frameworks have to be ad-hoc otherwise we'd have to edit the core code everytime we want to integrate something new.

But you're right, having a standard for PHP project management (maybe an addon to composer.json?), a standardised format of integration plugins, and a standardised format of indexing PHP code (I was thinking, SQLite db + ast nodes?) would benefit all projects greatly.

**Comment by cweagans** [comment]
> After all, code for integrating different CMS / frameworks have to be ad-hoc otherwise we'd have to edit the core code everytime we want to integrate something new.

That's kind of the idea - editing core code for integrating something new, I mean. It sounds kind of weird, but you can structure things in a modular way in core so that it's not a big tangled mess of spaghetti. It's worth it to do that, IMO, and then just name maintainers for individual integrations and give them access to the project so that they can merge PRs to their integrations (and submit their own code).

That way, interested parties can take ownership over a given framework's integration and add things quickly, everything is centralized, and nothing becomes unmaintained. If a given framework maintainer isn't responsive on their fork, it'll just slowly die. If a framework maintainer isn't responsive in the main project, then after some time, we can start looking for someone else to maintain that framework.

The main point I'm trying to make is this:

* Don't put a mechanism in place for loading external plugins.
* If someone wants to write a plugin, direct them toward adding support to core in a reasonable way.
* If there's a framework that's not supported by core and someone wants it, it *should* be supported by core (and not an external plugin).

Using @mkusher as an example again, the Symfony DI plugin he wrote for padawan could easily be merged into Padawan. To prevent all plugins from doing work all the time, a simple PluginInterface can be defined with a `pluginApplies()` method. If `pluginApplies()` returns FALSE for the current project, then none of the other code should be loaded or executed.

> But you're right, having [...] a standardised format of indexing PHP code (I was thinking, SQLite db + ast nodes?) would benefit all projects greatly.

I looked at sqlite for cweagans/theforce. It would be a good solution, but I wanted to have my project load the entire index into memory for performance. SQLite supports in-memory databases, which is *really* nice, but moving an on-disk database to an in-memory database and back is a tricky problem to solve in a reliable way. In any case, an on-disk SQLite database would be a good starting point.

**Comment by sekjun9878** [comment]
@cweagans Ah, I get what you mean.

**Comment by cweagans** [comment]
Couple more points:

* I really like @mkusher's idea of having the intelligence engine and the completer/linter/whatever as separate things. A common base that other tooling can be built on top of is a really great idea and could open the door for a wide ecosystem of intelligent PHP tooling like the Go community already has.
* To further clarify my points above, the framework/cms-specific plugins should live in the intelligence engine. The part that provides a way for editors to get autocomplete suggestions doesn't need to know about any of that.
* If somebody really, *really* needs to customize core in a way that is inappropriate for inclusion in core, then they can just add the intelligence engine as a dependency in their composer.json and extend the appropriate class. There would be some code duplication, but I don't think that's really something to worry about. We should encourage people to move things into core as much as possible.

**Comment by mkusher** [comment]
@cweagans I can't agree. SymfonyPlugin or PadawanDI plugin are just proof-of-concept, I'll break interface soon to support only project-based plugins(which will be listed in [padawan's project file](https://github.com/mkusher/padawan.php/issues/20)) and then will rewrite them. 
I don't understand why padawan's/phpcomplete's/phproject's or some_other_cool_lib_name's team should decide which plugins user should have and which he shouldn't. 
And for some closed-source projects developers should have easy and documented way to write plugins for their systems without hacking core lib source.
Look at idea's symfony2-plugin https://github.com/Haehnchen/idea-php-symfony2-plugin/tree/master/src/fr/adrienbrault/idea/symfony2plugin/ it is not so simple, there are yml/xml parsing and so, and so

**Comment by cweagans** [comment]
@mkusher The point is more that every person using the Symfony container component (no matter through what project) should get equally good autocomplete. That won't happen if we have Symfony Full Stack Framework plugins + Drupal 8 plugins + Laravel plugins, etc. We shouldn't decide what a user *shouldn't* have. We should decide that all users should have all things. Some of the plugins only apply in some contexts though (i.e. If I'm not using the Symfony container component, I don't want Symfony container completion suggestions), so plugins should have some way to detect whether or not the context is appropriate for them to jump in. If a plugin is not detecting the context properly with a given project, then that's a bug that should be fixed in core, and not a situation where someone should go off and write their own specific plugin for it.

Your point about closed source is valid, but here's an alternate way to solve that: closed source dev require's the intelligence engine and overrides the class that decides what plugins are included and simply adds his own. Really, though, the number of cases where that should be necessary should be almost nonexistent. The vast majority of PHPstorm's autocomplete functionality is included with the IDE, and the framework-specific plugins only add some minor tweaks (i.e. the Drupal plugin provides hook completions, for example). PHPstorm manages to provide decent autocompletion for closed-source projects with no plugins. Why can't we?

I'm not suggesting that we just dump everything into one class or something. The core architecture should be pretty modular, but if core doesn't handle > 95% of things OOTB, phpcomplete isn't going to get used that much ("It's buggy and not worth the trouble...I'll just use phpstorm"). If it DOES provide > 95% of the functionality that devs need, then why do we need plugins? I suppose it really depends on what the remaining 5% is, but having worked on many, *many* projects, both closed and open source, I know of precisely zero cases where I'd need to modify the behavior of the autocomplete tool that I'm using, and that's exactly how it should be.

Regarding https://github.com/Haehnchen/idea-php-symfony2-plugin, that's overly complicated because it's written in Java :) Symfony support should be broken down into many smaller pieces, though. For instance, having intelligence about the services in symfony/container would be one small bit of intelligence. There's some complexity, because those services can be added in PHP, in a YAML file, or in an XML file (and I'm sure there's other crazy things that people are doing), but that's a very concrete task. That can then be expanded on a bit for Drupal 8 container support, which uses the Symfony container, but in a different way (each module has a services.yml, so locating all of those and feeding them into the Symfony container would be the main responsibility). Actual code completion for BOTH of those frameworks, though, doesn't require anything special. If the engine can gather data from global functions and classes and understand scoping, that gets you 80-90% of the way there with pretty much any PHP framework.

**Comment by cweagans** [comment]
Regarding distribution, if phpcomplete is packaged as a docker container, we can ship it with php7 + the ast extension, which makes a lot of this much easier and faster.

**Comment by mkusher** [comment]
@cweagans php-di uses `DI` namespace for theirs' container. Some self-written containers may also use `DI` namespace, but have completely different API. Or they will have partial matching to php-di api. My point is that user should be able to turn on plugins for any of his project and turn them off. By default it should understand plain php as much as possible, but framework specific things should go in plugins.
Of course, since symfony2 is de-facto standard in php world we can include this plugin in default distribution, but not in the core, I think.

**Comment by cweagans** [comment]
I'd also suggest that https://youtrack.jetbrains.com/issues?q=project%3A+PhpStorm+%23Fixed+ is a good place to look for weird edge cases that a php completion engine will need to work through. PHPstorm has a very large head start on intelligent code completion and an entire staff of people working on it, but we can at least see the history of bugs that they've fixed, so that's a plus.

phpcomplete should also use https://github.com/JetBrains/phpstorm-stubs for completion/docblock info for built-in functions/classes/whatever.

**Comment by mkusher** [comment]
@cweagans padawan.php uses phpstorm-stubs, but with a few fixes

**Comment by cweagans** [comment]
@mkusher Why should I have to go download a plugin to get decent support for Symfony? It should just work. Ditto with any other open source component. Regardless of whether or not there's a 3rd party plugin ecosystem, there's no need for users to be able to manually toggle a plugin on or off. If there's a Symfony container in the project, the Symfony container plugin should kick in and do the right thing. If there's a php-di container in the project, that plugin should activate when the project is indexed.

Maybe - *maybe* - a custom DI container would need a plugin written to work properly, but again:

* Write the appropriate class
* Extend the main completion engine class in phpcomplete
* Call `$engine->addPlugin('Fully\Qualified\Class');`

This way, people can still add functionality specific to their own closed source project, the core code is very comprehensive and doesn't require a million plugins to achieve parity with phpstorm's completion, and autocomplete for custom projects gets versioned (almost by necessity) with the project code, which is exactly where it should live IMO.

**Comment by mkusher** [comment]
@cweagans It is something like single responsibility principle. I think core team should improve plain php completion, because I'm sure that there are a lot of work to be done. And symfony support needs also a lot of work to be done, for example <2.7 symfony and 2.8 symfony `app/console container:debug` has different output format.
We can distribute padawan(or phpcomplete) core with few plugins, like symfony/drupal or so. And I really see no profit in monolithic architecture

**Comment by mkusher** [comment]
Some time ago @complex857 sent me https://github.com/lvht/phpcd.vim by @lvht but I just saw a video in README

**Comment by cweagans** [comment]
@mkusher As a compromise, what would you think about including most (if not all) plugins with core OOTB so that there's a centralized place for people to work on code intelligence for different open source frameworks/libs/whatever, and then allowing for third party plugins by requiring that the plugins use the composer [type](https://getcomposer.org/doc/04-schema.md#type) of 'phpcomplete-plugin' (or something like that) and automatically registering plugins provided by phpcomplete-plugin packages in a given project.

If it's absolutely necessary to allow users a way to disable a specific plugin through an optional config file, I guess I'd be okay with that, but I still feel very strongly that we should just let the individual plugins decide when they apply to a given context, as that can be determined programmatically very easily in almost any case (I can't think of one, but I won't rule out a case where that's not possible).

**Comment by cweagans** [comment]
Yet another approach is to not have a separate project and instead build this functionality into srclib (https://github.com/sourcegraph/srclib, and specifically, https://github.com/sourcegraph/srclib-php).

**Comment by mkusher** [comment]
> I still feel very strongly that we should just let the individual plugins decide when they apply to a given context
I think user knows better whether he'd like to have Symfony/Drupal/Silex/PHP-DI/Pimple/OtherSuperbDI and which options pass to them for their projects. Behat doesn't try to guess whether it should use Symfony plugin or not

**Comment by cweagans** [comment]
Okay, give him a way to override it, but 99% of the time, we can just look at what's installed with Composer. If `symfony/dependency-injection` is in the list of packages, there's a really good chance that the user is going to want completions for the symfony container services. I'm not sure what behat has to do with this.

It shouldn't be necessary to configure a bunch of crap just to get autocompletion working, and there's no need to do it in most cases. It should Just Work&trade;.

**Comment by mkusher** [comment]
ok, we have completely different ideologies :)
I think that user should control most of the things and system can be configured in different ways by user and you think that system should be configured by itself without a user. And the truth is somewhere between :)

**Comment by cweagans** [comment]
You misunderstand. I think things should have a reasonable default config that works in most cases. If a user wants to tweak that config, more power to him. For this particular instance, I don't think it's *necessary* to tweak the config because the system can do the right thing without user input in most cases, but I concede that the user may need to configure things sometimes.

What I disagree with is the idea that in order to get completion working, I need to 1) go download a plugin, 2) configure the plugin, 3) test to make sure I've configured it right, 4) reconfigure things based on step 3, 5) do work that matters (as opposed to 1) download/start phpcomplete 2) start working).

**Comment by mkusher** [comment]
I'm thinking about current state of php code tools like about php web tools before symfony2. Now we need standard pack of components, that any code tool can use. php won a lot with symfony2 web(and not only web) components and I think php project/code tools components may help php world to make next step in ecosystem evolution.

**Comment by cweagans** [comment]
Understood. I guess I'm thinking more about the assembled distribution - more from the perspective of the user of the tool. If each individual component is configurable (like Symfony is), but they can be assembled into a more opinionated tool (again, like Symfony - Laravel is a good example), and the tool is what gets widely used, I'd be okay with that.

**Comment by mkusher** [comment]
@cweagans it also like symfony2 has: there are a lot of components and there is symfony2 standard edition

**Comment by cweagans** [comment]
Exactly. Okay, so it sounds like a path forward here is:

* A code intelligence library. This would need to be configured, and is largely an API for other tooling to use.
* A collection of plugins that enhance the library for various frameworks/libraries/whatever. Not really a public API - just expands the completion suggestions that the code intelligence lib can provide.
* A code completion server that uses the lib and a collection of plugins. This would need less (if any) configuration, and should Just Work OOTB.
* Other tooling that depends on the code intelligence library can come later.

Is that right?

Some other ideas for tooling that aren't really filled by open source tools right now:

* A good code generator, and by "good", I mean largely what's in PHPStorm. User defined templates, framework specific actions, etc.
* Refactoring tools - again, modeled after what's in phpstorm. Detection of places where code can be simplified or generally done in a better way, pushing a method to a parent class (or overriding it in a child class), etc.
* Scope-aware code formatting and modeled after phpstorm.

Generally, I'd really love to have something like https://github.com/fatih/vim-go, but for PHP.

**Comment by sergeyz** [comment]
Nice to see a lot of feedback! Sorry for over-quoting, just want not to lose context

> I'd love to see this happen. This is a need I would like filled for Sublime Text.

More use cases we have - more flexible tool we can build, welcome!

> I'd be interested in contributing to this in my spare(?!) time.
> I likely won't have a lot of time to spend on this, but I'll follow the project and contribute where I can.

Sure. I think it will not take a lot of time for first time. 
I going to work on background in nearest future and some kind of spec. So comments and use cases are welcome.

> After doing a very significant amount of research on the topic of intelligent code completion, I've concluded that Omnisharp is the model that pretty much any code intelligence engine should follow. I'd also say that one should not attempt to integrate with each individual editor. Instead, integrate your completion engine with valloric/ycmd, and then you can more or less instantly leverage the ycmd integrations that exist for many editors already.

I follow your progress and really appreciate what you did. Totally agree with editors integration. 

> I'd also suggest taking a look at codeintel, which is the code intelligence component of Komodo Edit (open source). You can do pip install codeintel, and then run codeintel, and you get the Komodo code completion engine running as a standalone daemon. Among other languages, codeintel supports PHP reasonably well. I think the sources for that are here, but I'm not sure.

I agree that both ycmd and codeintel quite useful tools and codeintel best choice for now. But, using them can lead us to bottlenecks hard to solve. They are:

- Framework integration - Currently codeintel doesn't support it in way we can use it widely.
- Extendability - Include integration into ycmd core can lead us to case when we will not able to solve issue efficiently. Another disadvantage is Python, supporting it can become mess, and those one looking for PHP code completion will have php environment in most cases, but I'm not sure about Python

> Oh, one other thing: I think for something like this, a more monolithic model makes sense. That is, rather than having a plugin system that allows end-users to customize what the completer will and won't do, it should just work more like the completion engine that's built into PHPstorm. It should handle anything you throw at it without having to fiddle with a config file.

Its totally on tools side. It can provide basic working configuration that you don't need to reconfigure, and that is common case. But same time it should give an opportunity to change completion to fit developer requirements

> I'm currently working on refactoring padawan.php, because have many ideas, how I can improve type guessing, project management and index updating. Have a look at new structure I'm working on https://github.com/mkusher/padawan.php/tree/features/custom-scopes/src

Sure, will check it, thanks!

> What, I think, can be unified - classes and interfaces, that store information about project, classes, interfaces, methods and etc. They can be used for completion, for linter, for refactoring and so.
I even started to think that padawan should be a lib for a php-project management and other things like completer or linter should be plugins.

It will look like some kind of framework, I don't have plans to do same. One usecase I have is that we can use index to create `Find usages` functionality. But I think that is not core functionality and should be done by tool based on current one.

> I haven't tried this personally, but a "gold-standard" open source implementation of PHP completion may be available at Valloric/ycmd#120. Of course, the aim of this project is to possibly port that functionality in to PHP code so PHP users can maintain it, but just pointing that out that according to the author that patch is fully functional and ready to roll.

Actually I thought to port codeintel to php, but not sure we can support bugfixes and features on same level with codeintel. Going to use another way, I will add separate comment with ideas

> Yes, I suppose that's true. However, those are just enhancements. Why not build that directly into the completer and have the completer recognize when there's a symfony project or a Drupal project or something? I guess it doesn't really matter if it's through a plugin, but the thing that I'm concerned with is discoverability. It's hard enough to find a decent PHP completion engine. It's even harder to find plugins for it.

I don't have clear vision on this. We can't support all frameworks as core functionality, just think over versioning support on Symfony example. In case we will include all framework integration to core I'd like to see frameworks with top priory integration and support

> A set of core included-in-default-distrubution plugins can easily solve that issue. That's more of a problem of distribution and code management rather than an architectural decision isn't it? After all, code for integrating different CMS / frameworks have to be ad-hoc otherwise we'd have to edit the core code everytime we want to integrate something new.

Agree, in addition I would like to say that we can easily postpone distribution question until codebase is not so large. Its a good option when someone wants to use part of code, but not a silver bullet to do it from the beginning

> Don't put a mechanism in place for loading external plugins.

Think it should be available when there is no need to maintain some integration anymore. Just an option but in result we can declare maintainers responsibility over tool core. All integrations in core are always supported. Anyway, there is composer, so I think distribution is not a problem

**Comment by sergeyz** [comment]
@mkusher @cweagans 

I think I'm going to agree with both of you. We should have both ways - default configured distribution and possibility to configure your own one

As for threads - github is not so good with that, I'm going to give an access for you , feel free to create Issues and split discussions into smaller ones. Just to make things clear

**Comment by sergeyz** [comment]
> Yet another approach is to not have a separate project and instead build this functionality into srclib (https://github.com/sourcegraph/srclib, and specifically, https://github.com/sourcegraph/srclib-php).

My next comment was about engine actually, I was looking into https://github.com/JetBrains/Grammar-Kit and BNF stuff, but looks like I need some time to investigate srclib

**Comment by sekjun9878** [comment]
Is it possible to have this "all-around PHP intelligence project" named something other than phpcomplete? I think one criticial functionality that all PHP "code-intelligence" projects are missing right now is implicit type inference and checking. I find that it's usually not autocomplete that stops me (when not using PHPStorm) but coding without the guarantees that PHPStorm gives me.

Naming it phpcomplete would unnecessarily restrict it to the autocomplete domain only. Personally, I have had "Elphp" in mind. It's short, concise, comes from Elephpant that every PHP user knows, but I may be biased there because that's the name of project I'm already working on.

As for the engine, I believe it should be purely based on PHP, and depend only on standard PHP extensions. One thing we have in common is that we are PHP coders first and foremost.

**Comment by cweagans** [comment]
> As for the engine, I believe it should be purely based on PHP, and depend only on standard PHP extensions. One thing we have in common is that we are PHP coders first and foremost.

On one hand, I agree with you because that allows developers to improve their own tools, but on the other hand, PHP is really slow for things like this. If we can offload pieces of work to faster implementations (e.g. php extensions like the ast extension for php 7), the tool works better...

**Comment by sekjun9878** [comment]
@cweagans PHP extension seems fine, as long as we can make it available on apt and rpm, so we don't have to recompile PHP to use it.

**Comment by cweagans** [comment]
> PHP extension seems fine, as long as we can make it available on apt and rpm, so we don't have to recompile PHP to use it.

Even better if there's a userspace fallback. For instance, we could use the ast extension if it's available, and if it's not, then fall back to https://github.com/nikic/PHP-Parser. Not sure if the output formats are similar enough to do that in a reasonable way, though. Given that the ast extension is for PHP 7, though, it's very unlikely that the ast extension will be installable from anywhere for a long while. It's not even in homebrew (for Mac) yet. Installing the extension is pretty easy, though. As I mentioned before, we could even ship a Docker container that has everything needed already installed.

> Naming it phpcomplete would unnecessarily restrict it to the autocomplete domain only. Personally, I have had "Elphp" in mind. It's short, concise, comes from Elephpant that every PHP user knows, but I may be biased there because that's the name of project I'm already working on. Edit: Preferrably the project name should be < 5 characters for easy usage in command line.

The discussion above was talking about a wider set of tools and components. I'm generally +1 to renaming the org `phptools` or something more generic, and then having a bunch of small projects inside that org for the intelligence engine, a completion server, framework plugins, etc. Don't much care about the length of the name of the completion daemon. Tab completion and shell aliases make this a nonissue in my book.

**Comment by sekjun9878** [comment]
@cweagans Is there really a performance problem with PHP-Parser? For what I've tried it on it was pretty fast, and results (the index) can be cached.

I was thinking that most of the refactoring tools etc. should be available via command line. A short name benefits that usage. In addition, it could be used also instead of the php command to introduce dynamic type checks before running the code (similar to mypy-lang.org). But yeah, a static type analyser is a gem that's missing in my opinion (there's scrutinizer-ci, but it's hosted online and closed source).

**Comment by cweagans** [comment]
@sekjun9878 Well, I observed one in cweagans/theforce. Parsing a single file seems pretty fast, but when you have a large project (like Drupal 8, which has 1000s of files), it's observably slow. More to the point, though, the ast extension may be more accurate. Since PHP7 uses an AST internally, using that data will be more representative of how the interpreter will view things. I could go either way though. There's pros and cons with both approaches.

**Comment by sekjun9878** [comment]
I just found that PHPStorm also has an offline inspection script which may perform type checkng? https://www.jetbrains.com/phpstorm/help/running-inspections-offline.html

**Comment by sekjun9878** [comment]
srclib sounds promising. Right now srclib doesn't support autocompletion, but I've talked with them on their slack channel about having it on their roadmap for their next release. The code analysis base is already there though, along with editor integration provided from them (open source).

They seem pretty interested in working with us for PHP support. But I'm worried using them as a base would make us locked into them, and therefore their company product of SourceGraph?

It seems that we do get total control over our domain though, so integration with Symfony2, and other things can be done on our side easily.

**Comment by cweagans** [comment]
Having the common language analysis base in place is a pretty big selling point, IMO.

Going this route means that:

1. We don't have to build the really low level stuff for storing code intelligence data, since that's provided by srclib, so we really only need to focus on extracting useful information from PHP projects and handing it off to srclib.
2. Tools that integrate with srclib will automatically get support for PHP when/if we get the PHP support working.
3. The improvements that we contribute via PHP support *would* increase their commercial product's value, but I think I'm okay with that if everyone else is on board.
4. We could build a really light autocomplete daemon in Go that uses srclib and all of the language toolchains to provide autocomplete suggestions. I'm happy to take this on if nobody else wants to. I'm generally -1 to writing long running processes in PHP if it's avoidable, and Go is much easier to deal with in that regard.
5. We can build refactoring tools on top of srclib that work with any language (even if we only focus on testing with PHP).
6. We can build linting/formatting tools on top of srclib that work with any language (again, even if we're only focusing on PHP).

I think this is a pretty great opportunity. The only thing we need to resolve is how we're going to handle the plugin setup. Building the intelligence features into srclib changes things a bit, given that (as far as I'm aware) the PHP toolchain would need to just handle things intelligently. I'm not sure how to build plugins for a srclib plugin, but that might be something that we can work with the srclib guys on...

**Comment by sergeyz** [comment]
7. We can build static analysis tools or provide some basic features (index)

**Comment by sekjun9878** [comment]
@cweagans srclib looks like it makes it easy to build a language integration https://github.com/sourcegraph/srclib/blob/master/docs/sources/api/data-model.md. I'm not sure how the grapher output is supposed to be, but I'm sure can ask on their slack. We build our own intelligence engine (similar to pysonar and rubysonar, and padawan.php) and we can 1. hand off the data to srclib 2. if srclib does not suit us, we can build our raw tooling to use that format e.g. for refactoring.

I guess our first priority should be to build something like pysonar, and not just something like phpctags if we want intelligent code inspections, type checking, refactoring and autocompletion in the future.

Is the goal of phpcomplete simply autocompletion, or something more advanced like omnisharp?

**Comment by cweagans** [comment]
Autocomplete, refactoring, and type inference all start with extracting information from the project's codebase. If we can't do that (extract info), then the rest is all a moot point :)

I'd love to get it to the point where we have a lot of information in srclib, and then we can start building tools on top of srclib to use/interpret that data in a meaningful way. Autocomplete, typechecking, refactoring, etc. Basically, if we can provide all of the omnisharp features for PHP, I'd be happy.

I wish we had something like NRefactory for PHP, because that would make this a lot easier.

**Comment by sekjun9878** [comment]
Can we set up an IRC channel on Freenode because this thread is starting to get quite long @sergeyz?

**Comment by mkusher** [comment]
I think gitter will also work well

**Comment by sergeyz** [comment]
Done https://gitter.im/phpcomplete/phpcomplete

**Comment by lvht** [comment]
Just another faster and smarter php complete plugin [PHPCD](https://github.com/phpvim/phpcd.vim/blob/master/README_en.md)

**Comment by sergeyz** [comment]
https://github.com/felixfbecker/php-language-server
https://github.com/Microsoft/language-server-protocol
https://github.com/Microsoft/tolerant-php-parser
